gossip/tx_scrambler_test.go

```diff
@@ -0,0 +1,645 @@
+					nonce:  1,
+					gas:    5,
+				},
+			},
+		},
+		{
+			name: "repeated addresses, nonces and gas",
+			entries: []*scramblerEntry{
+				{
+					hash:   common.Hash{1},
+					sender: common.Address{1},
+					nonce:  1,
+					gas:    1,
+				},
+				{
+					hash:   common.Hash{2},
+					sender: common.Address{2},
+					nonce:  2,
+					gas:    2,
+				},
+				{
+					hash:   common.Hash{3},
+					sender: common.Address{3},
+					nonce:  3,
+					gas:    3,
+				},
+				{
+					hash:   common.Hash{4},
+					sender: common.Address{2},
+					nonce:  2,
+					gas:    2,
+				},
+				{
+					hash:   common.Hash{5},
+					sender: common.Address{1},
+					nonce:  1,
+					gas:    1,
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			res1 := test.entries
+			res2 := deepCopyEntries(res1)
+			// shuffle one array
+			shuffleEntries(res2)
+
+			res1 = filterAndOrderTransactions(res1)
+			res2 = filterAndOrderTransactions(res2)
+			if !reflect.DeepEqual(res1, res2) {
+				t.Error("slices have different order - algorithm is not deterministic")
+			}
+		})
+	}
+}
+
+func TestTxScrambler_GetExecutionOrder_SortRemovesDuplicateHashes(t *testing.T) {
+	entries := []*scramblerEntry{
+		{hash: common.Hash{1}},
+		{hash: common.Hash{2}},
+		{hash: common.Hash{3}},
+		{hash: common.Hash{2}},
+		{hash: common.Hash{1}},
+	}
+	shuffleEntries(entries)
+	entries = filterAndOrderTransactions(entries)
+
+	checkDuplicateHashes(t, entries)
+}
+
+func TestTxScrambler_GetExecutionOrder_SortsSameSenderByNonceAndGas(t *testing.T) {
+	entries := []*scramblerEntry{
+		{
+			hash:   common.Hash{1},
+			sender: common.Address{1},
+			nonce:  1,
+		},
+		{
+			hash:   common.Hash{2},
+			sender: common.Address{1},
+			nonce:  2,
+		},
+		{
+			hash:   common.Hash{3},
+			sender: common.Address{1},
+			nonce:  3,
+			gas:    1,
+		},
+		{
+			hash:   common.Hash{4},
+			sender: common.Address{1},
+			nonce:  3,
+			gas:    2,
+		},
+		{
+			hash:   common.Hash{5},
+			sender: common.Address{1},
+			nonce:  4,
+		},
+	}
+	shuffleEntries(entries)
+	entries = filterAndOrderTransactions(entries)
+
+	for i := 0; i < len(entries); i++ {
+		for j := i + 1; j < len(entries); j++ {
+			if entries[i].sender == entries[j].sender {
+				if entries[i].nonce > entries[j].nonce {
+					t.Errorf("incorrect nonce order %d must be before %d", entries[j].nonce, entries[i].nonce)
+				}
+				if entries[i].nonce == entries[j].nonce && entries[i].gas < entries[j].gas {
+					t.Errorf("incorrect gas order %d must be before %d", entries[j].gas, entries[i].gas)
+				}
+			}
+		}
+	}
+}
+
+func TestTxScrambler_GetExecutionOrder_RandomInput(t *testing.T) {
+	// this tests these input sizes:
+	// 1, 4, 16, 64, 256, 1024
+	for i := int64(1); i <= 1024; i = i * 4 {
+		input := createRandomScramblerTestInput(i)
+		cpy := deepCopyEntries(input)
+		shuffleEntries(cpy)
+		input = filterAndOrderTransactions(input)
+		cpy = filterAndOrderTransactions(input)
```

Comment:
Should this get `cpy` as an argument?
```suggestion
		cpy = filterAndOrderTransactions(cpy)
```